src/Appwrite/Platform/Workers/Messaging.php

```diff
@@ -60,48 +66,314 @@ public function action(Message $message): void
             return;
         }
 
-        if (empty($payload['recipient'])) {
-            Console::error('Recipient arg not found');
-            return;
+        if (!\is_null($payload['message']) && !\is_null($payload['recipients'])) {
+            if ($payload['providerType'] === 'SMS') {
+                $this->processInternalSMSMessage(new Document($payload['message']), $payload['recipients']);
+            }
+        } else {
+            $message = $dbForProject->getDocument('messages', $payload['messageId']);
+
+            $this->processMessage($dbForProject, $message);
         }
+    }
 
-        if (empty($payload['message'])) {
-            Console::error('Message arg not found');
-            return;
+    private function processMessage(Database $dbForProject, Document $message): void
+    {
+        $topicsId = $message->getAttribute('topics', []);
+        $targetsId = $message->getAttribute('targets', []);
+        $usersId = $message->getAttribute('users', []);
+
+        /**
+        * @var Document[] $recipients
+        */
+        $recipients = [];
+
+        if (\count($topicsId) > 0) {
+            $topics = $dbForProject->find('topics', [Query::equal('$id', $topicsId)]);
+            foreach ($topics as $topic) {
+                $recipients = \array_merge($recipients, $topic->getAttribute('targets'));
+            }
         }
 
-        $sms =  match ($this->dsn->getHost()) {
-            'mock' => new Mock($this->user, $this->secret), // used for tests
-            'twilio' => new Twilio($this->user, $this->secret),
-            'text-magic' => new TextMagic($this->user, $this->secret),
-            'telesign' => new Telesign($this->user, $this->secret),
-            'msg91' => new Msg91($this->user, $this->secret),
-            'vonage' => new Vonage($this->user, $this->secret),
-            default => null
-        };
+        if (\count($usersId) > 0) {
+            $users = $dbForProject->find('users', [Query::equal('$id', $usersId)]);
+            foreach ($users as $user) {
+                $recipients = \array_merge($recipients, $user->getAttribute('targets'));
+            }
+        }
 
-        if (empty(App::getEnv('_APP_SMS_PROVIDER'))) {
-            Console::error('Skipped sms processing. No Phone provider has been set.');
-            return;
+        if (\count($targetsId) > 0) {
+            $targets = $dbForProject->find('targets', [Query::equal('$id', $targetsId)]);
+            $recipients = \array_merge($recipients, $targets);
         }
 
-        $from = App::getEnv('_APP_SMS_FROM');
+        $internalProvider = $dbForProject->findOne('providers', [
+            Query::equal('enabled', [true]),
+            Query::equal('type', [$recipients[0]->getAttribute('providerType')]),
+        ]);
```

Comment:
@fanatic75, what if `findOne()` doesn't find anything?

Reply:
this is fixed